lib/Infer/ExhaustiveSynthesis.cpp

```diff
@@ -0,0 +1,568 @@
+
+    for (auto I : Guesses) {
+      // separate sub-expressions by copying vars
+      std::map<Inst *, Inst *> InstCache;
+      std::map<Block *, Block *> BlockCache;
+
+      Inst *Eq = IC.getInst(Inst::Eq, 1, {getInstCopy(LHS, IC, InstCache, BlockCache, 0, true),
+                                          getInstCopy(I, IC, InstCache, BlockCache, 0, true)});
+
+      Ante = IC.getInst(Inst::And, 1, {Ante, Eq});
+      separateBlockPCs(BPCs, BPCsCopy, InstCache, BlockCache, IC, 0, true);
+      separatePCs(PCs, PCsCopy, InstCache, BlockCache, IC, 0, true);
+    }
+
+    if (DebugLevel > 2) {
+      llvm::errs() << "\n\n--------------------------------------------\nBigQuery\n";
+      ReplacementContext RC;
+      RC.printInst(Ante, llvm::errs(), false);
+    }
+
+    // (LHS != i_1) && (LHS != i_2) && ... && (LHS != i_n) == true
+    InstMapping Mapping(Ante, IC.getConst(APInt(1, true)));
+    std::string Query = BuildQuery(IC, BPCsCopy, PCsCopy, Mapping, 0, /*Negate=*/false);
+    if (Query.empty())
+      return std::make_error_code(std::errc::value_too_large);
+    bool BigQueryIsSat;
+    EC = SMTSolver->isSatisfiable(Query, BigQueryIsSat, 0, 0, Timeout);
+    if (EC)
+      return EC;
+    if (!BigQueryIsSat) {
+      if (DebugLevel > 2)
+        llvm::errs() << "big query is unsat, all done\n";
+      return EC;
+    } else {
+      if (DebugLevel > 2)
+        llvm::errs() << "big query is sat, looking for small queries\n";
+    }
+  }
+  // find the valid one
+  int Unsat = 0;
+  int GuessIndex = -1;
+  for (auto I : Guesses) {
+    GuessIndex++;
+    if (DebugLevel > 2) {
+      llvm::errs() << "\n\n--------------------------------------------\nguess " << GuessIndex << "\n\n";
+    }
+
+    std::vector<Inst *> ConstList;
+    hasConstant(I, ConstList);
+
+    std::map<Inst*, std::vector<llvm::APInt>> BadConsts;
+    int Tries = 0;
+
+  again:
+    if (Tries > 0 && DebugLevel > 2)
+      llvm::errs() << "\n\nagain:\n";
+
+    // this SAT query will give us possible constants
+
+    // avoid choices for constants that have not worked out in previous iterations
+    // ((R1 != C11 ) \/ (R2 != C21 )) /\ ((R1 != C12 ) \/ (R2 != C22 )) /\ ...
+    Inst *AvoidConsts = IC.getConst(APInt(1, true));
+    if (!BadConsts.empty()) {
+      for (unsigned i = 0; i < BadConsts[ConstList[0]].size(); ++i) {
+        Inst *Ante = IC.getConst(APInt(1, false));
+        for (auto C : ConstList) {
+          Inst *Ne = IC.getInst(Inst::Ne, 1, {IC.getConst(BadConsts[C][i]), C });
+          Ante = IC.getInst(Inst::Or, 1, {Ante, Ne});
+        }
+        AvoidConsts = IC.getInst(Inst::And, 1, {Ante, AvoidConsts});
+      }
+    }
+
+    std::vector<Inst *> Vars;
+    findVars(LHS, Vars);
+
+    std::map<Inst *, std::vector<llvm::APInt>> TriedVars;
+    std::map<Inst *, llvm::APInt> ConstMap;
+
+    {
+      Inst *Ante = IC.getConst(APInt(1, true));
+      if (!Vars.empty()) {
+        // Try two combinations of values at a time
```

Comment:
can you explain this more? where does 2 come from?

Reply:
Currently this value can be set to any non-negative number. But since the current implementation will not specialize a input variable with two same value. If some program has input type (i1, i32), since the type of the first argument is i1, so there will be only two specialized input combinations, like (0, C1) and (1, C2).

Reply:
The fix would be introducing some algorithm which can produce (0, C_1) (1, C_2), (0, C_3), (1, C_4) ...

Reply:
Can you write a comment about this, and also make the 2 into a named constant defined at the top of the file?